src/merkle/simple_smt/mod.rs

```diff
@@ -66,36 +60,47 @@ impl SimpleSmt {
         })
     }
 
-    /// Appends the provided entries as leaves of the tree.
+    /// Returns a new [SimpleSmt] instantiated with the specified depth and with leaves
+    /// set as specified by the provided entries.
     ///
-    /// # Errors
+    /// All leaves omitted from the entries list are set to [ZERO; 4].
     ///
-    /// The function will fail if the provided entries count exceed the maximum tree capacity, that
-    /// is `2^{depth}`.
-    pub fn with_leaves<R, I>(mut self, entries: R) -> Result<Self, MerkleError>
+    /// # Errors
+    /// Returns an error if:
+    /// - If the depth is 0 or is greater than 64.
+    /// - The number of entries exceeds the maximum tree capacity, that is 2^{depth}.
+    /// - The provided entries contain multiple values for the same key.
+    pub fn with_leaves<R, I>(depth: u8, entries: R) -> Result<Self, MerkleError>
```

Comment:
We might be creating a bit of redundancy here. These builder functions, often, takes a single argument to incorporate into the type. The proposed implementation might overlap with `Self::new`, where both kinda exercise the same function.

Reply:
Yeah - `with_leaves()` internally just calls `new()` and then updates it. The main reason for this change is to simplify usage of the external interface. Previously, we'd have do something like this:
```Rust
let smt = SimpleSmt::new(64)
    .expect("depth is good")
    .with_leaves(leaves)
    .expect("failed to create tree").
```
Now, this can be done as:
```Rust
let smt = SimpleSmt::with_leaves(64, leaves).expect("failed to create tree").
```